CppSamples/EditData/SnapGeometryEditsWithRules/SnapGeometryEditsWithRules.cpp

```diff
@@ -0,0 +1,559 @@
+    emit assetTypeChanged(utilityEle->assetType()->name());
+    emit assetGroupChanged(utilityEle->assetGroup()->name());
+    emit isElementSelectedChanged();
+
+    setSnapSettings(utilityEle->assetType());
+  });
+}
+
+void SnapGeometryEditsWithRules::setMapView(MapQuickView* mapView)
+{
+  if (m_mapView == mapView)
+  {
+      return;
+  }
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+  m_mapView->setGeometryEditor(new GeometryEditor(this));
+#if defined(Q_OS_IOS) || defined(Q_OS_ANDROID)
+  m_mapView->geometryEditor()->setTool(new ReticleVertexTool(this));
+#endif
+
+  auto* snapSettings = new SnapSettings(this);
+  snapSettings->setEnabled(true);
+  snapSettings->setFeatureSnappingEnabled(true);
+  m_mapView->geometryEditor()->setSnapSettings(snapSettings);
+
+  auto* graphicsOverlay = new GraphicsOverlay(this);
+  graphicsOverlay->setRenderer(m_defaultGraphicsOverlayRenderer);
+
+  const QString graphicJson = R"({"paths":[[[-9811826.6810284462,5132074.7700250093],[-9811786.4643617794,5132440.9533583419],[-9811384.2976951133,5132354.1700250087],[-9810372.5310284477,5132360.5200250093],[-9810353.4810284469,5132066.3033583425]]],"spatialReference":{"wkid":102100,"latestWkid":3857}})";
+  graphicsOverlay->graphics()->append(new Graphic(Geometry::fromJson(graphicJson), this));
+
+  m_mapView->graphicsOverlays()->append(graphicsOverlay);
+
+  connect(m_mapView, &MapQuickView::mouseClicked, this, &SnapGeometryEditsWithRules::onMapViewClicked);
+
+  emit mapViewChanged();
+}
+
+MapQuickView* SnapGeometryEditsWithRules::mapView() const
+{
+  return m_mapView;
+}
+
+bool SnapGeometryEditsWithRules::isElementSelected() const
+{
+  return (m_selectedFeature != nullptr);
+}
+
+SnapSourceListModel* SnapGeometryEditsWithRules::snapSourceListModel() const
+{
+  return m_snapSourceListModel;
+}
+
+bool SnapGeometryEditsWithRules::geometryEditorStarted() const
+{
+  return (m_mapView && m_mapView->geometryEditor() && m_mapView->geometryEditor()->isStarted());
+}
+
+void SnapGeometryEditsWithRules::startEditor()
+{
+  if (!m_selectedFeature || !m_selectedFeature->featureTable() || !dynamic_cast<FeatureLayer*>(m_selectedFeature->featureTable()->layer()))
+  {
+    return;
+  }
+
+  // Get the symbol for the selected feature.
+  Symbol* selectedFeatureSymbol = dynamic_cast<GeodatabaseFeatureTable*>(m_selectedFeature->featureTable())->layerInfo().drawingInfo().renderer()->symbol(m_selectedFeature);
+
+  // Set the vertex symbol for the geometry editor tool.
+  auto* geometryEditorStyle = m_mapView->geometryEditor()->tool()->style();
+  geometryEditorStyle->setVertexSymbol(selectedFeatureSymbol);
+  geometryEditorStyle->setFeedbackVertexSymbol(selectedFeatureSymbol);
+  geometryEditorStyle->setSelectedVertexSymbol(selectedFeatureSymbol);
+
+  // Hide the selected feature.
+  dynamic_cast<FeatureLayer*>(m_selectedFeature->featureTable()->layer())->setFeatureVisible(m_selectedFeature, false);
```

Comment:
Review throughout. I won't comment on them all, but it looks like static_cast will suffice for all these.